contracts/utils/OptimisticTimelock.sol

```diff
@@ -0,0 +1,156 @@
+    function deny(bytes32 txHash) external;
+    function execute(Call[] calldata functionCalls) external returns (bytes[] calldata results);
+    function executeRepeated(Call[] calldata functionCalls, uint256 salt) external returns (bytes[] calldata results);
+}
+
+/// @dev With the OptimisticTimelock, all proposals are automatically approved, but an account with `deny`
+/// permissions can take them off the queue.
+/// @notice In the event of a DoS attack, simply take the `schedule` permissions from the atacker.
+contract OptimisticTimelock is IOptimisticTimelock, AccessControl {
+    using IsContract for address;
+
+    enum STATE { UNKNOWN, SCHEDULED, DENIED }
+
+    struct Proposal {
+        STATE state;
+        uint32 eta;
+    }
+
+    uint32 public constant GRACE_PERIOD = 14 days;
+    uint32 public constant MINIMUM_DELAY = 2 days;
+    uint32 public constant MAXIMUM_DELAY = 30 days;
+
+    event DelaySet(uint256 indexed delay);
+    event Proposed(bytes32 indexed txHash, uint32 eta);
+    event Denied(bytes32 indexed txHash);
+    event Executed(bytes32 indexed txHash);
+
+    uint32 public delay;
+    mapping (bytes32 => Proposal) public proposals;
+
+    constructor(address governor) AccessControl() {
+        delay = 0; // delay is set to zero initially to allow testing and configuration. Set to a different value to go live.
+
+        // Each role in AccessControl.sol is a 1-of-n multisig. It is recommended that trusted individual accounts get `schedule`
+        // and `execute` permissions, while only the governor keeps `deny` permissions. The governor should keep the `schedule`
+        // and `execute` permissions, but use them only in emergency situations (such as all trusted individuals going rogue).
+        _grantRole(IOptimisticTimelock.schedule.selector, governor);
+        _grantRole(IOptimisticTimelock.scheduleRepeated.selector, governor);
+        _grantRole(IOptimisticTimelock.deny.selector, governor);
+        _grantRole(IOptimisticTimelock.execute.selector, governor);
+        _grantRole(IOptimisticTimelock.executeRepeated.selector, governor);
+
+        // Changing the delay must now be executed through this Timelock contract
+        _grantRole(IOptimisticTimelock.setDelay.selector, address(this)); // bytes4(keccak256("setDelay(uint256)"))
+
+        // Granting roles (schedule, deny, execute, setDelay) must now be executed through this Timelock contract
+        // For increased security, ROOT can be given to a non-optimistic Timelock and removed from `this`.
+        _grantRole(ROOT, address(this));
+        _revokeRole(ROOT, msg.sender);
+    }
+
+    /// @dev Change the delay for denyd proposals
+    function setDelay(uint32 delay_) external override auth {
+        require(delay_ >= MINIMUM_DELAY, "Must exceed minimum delay.");
+        require(delay_ <= MAXIMUM_DELAY, "Must not exceed maximum delay.");
```

Comment:
If you're willing to lose the specifity of the message, you can save up to 14gwei by combining these two require statements into one :D